hudi-integ-test/src/main/java/org/apache/hudi/integ/testsuite/dag/nodes/ValidateAsyncOperations.java

```diff
@@ -0,0 +1,119 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.hudi.integ.testsuite.dag.nodes;
+
+import org.apache.hudi.common.util.Option;
+import org.apache.hudi.integ.testsuite.configuration.DeltaConfig.Config;
+import org.apache.hudi.integ.testsuite.dag.ExecutionContext;
+
+import org.apache.hadoop.fs.FileStatus;
+import org.apache.hadoop.fs.FileSystem;
+import org.apache.hadoop.fs.LocatedFileStatus;
+import org.apache.hadoop.fs.Path;
+import org.apache.hadoop.fs.RemoteIterator;
+
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * Node to validate data set sanity like total file versions retained, has cleaning happened, has archival happened, etc.
+ */
+public class ValidateAsyncOperations extends DagNode<Option<String>> {
+
+  public ValidateAsyncOperations(Config config) {
+    this.config = config;
+  }
+
+  @Override
+  public void execute(ExecutionContext executionContext, int itrCount) throws Exception {
+    if (config.getExecuteIterationCount() == itrCount) {
+      try {
+        log.warn("Executing ValidateHoodieAsyncOperations node {} with target base path {} ", this.getName(),
+            executionContext.getHoodieTestSuiteWriter().getCfg().targetBasePath);
+        String hudiPath = executionContext.getHoodieTestSuiteWriter().getCfg().targetBasePath;
+
+        int maxCommitsRetailed = executionContext.getHoodieTestSuiteWriter().getWriteConfig() != null ?
+            executionContext.getHoodieTestSuiteWriter().getWriteConfig().getCleanerCommitsRetained() + 1 : 11;
+        FileSystem fs = new Path(hudiPath)
+            .getFileSystem(executionContext.getHoodieTestSuiteWriter().getConfiguration());
+        Map<String, Integer> fileIdCount = new HashMap<>();
+        int maxVal = 0;
+        RemoteIterator<LocatedFileStatus> itr = fs.listFiles(new Path(hudiPath), true);
+        while (itr.hasNext()) {
```

Comment:
Let's chat about all of this logic 1-1. This code is very difficult to read..may be there is a better way to achieve what you are trying

Reply:
sure.